File: src/go/problems/RemoveElements.go
// https://leetcode.com/problems/remove-linked-list-elements

package problems

func ReverseList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	var result *ListNode
	current := head

	for current != nil {
		next := current.Next
		current.Next = result
		result = current
		current = next
	}

	return result
}

File: src/python/problems/remove_elements.py
# https://leetcode.com/problems/remove-linked-list-elements/

from typing import Optional
from models.list_node import ListNode

class RemoveElements:
    def Solution(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        prev = head
        current = head
        while current:
            if current.val != val:
                prev = current
            else:
                prev.next = current.next

            current = current.next

        return head if not head or head.val != val else head.next


RemoveElements().Solution([ListNode(1, ListNode(2, ListNode(4))), ListNode(1, ListNode(3, ListNode(4))), ListNode(1, ListNode(2, ListNode(4))), ListNode(1, ListNode(3, ListNode(4)))])
